enhancements/update/upgrade-without-registry.md

```diff
@@ -0,0 +1,893 @@
+  ...
+```
+
+We explored several alternatives for the format of the bundle, and using a dump
+of a registry server to be the best one.
+
+A plain file copy is feasible only if the target directory is empty and nothing
+is using it. That is why we are suggesting to add full support to the
+additionalimagestores setting in storage.conf. If that was usable (via
+`ContainerRuntimeConfig`) then we could create such an additional directory, copy
+the files there and reload CRI-O. The size of the (uncompressed) bundle for
+this would be approximately 32 GiB, and it would be there for ever, at least
+till the next upgrade, because those additional directories are read only.
+
+It is also possible to copy the images using the equivalent of `skopeo copy
+containers-storage:... containers-storage: ...`. That doesn't need need the
+permanent additional directory, and doesn't need to shutdown or reload CRI-O,
+but it does need those 32 GiB temporarily, while the copy is in progress.
+
+Another possibility is to use the equivalent of skopeo copy `docker://...
+containers-storage:...`. That is where the temporary registry comes in place.
+The advantage is that the format used by the registry to store the images is
+more efficient: it only needs 16 GiB for a release. That reduces the size of
+the bundle and the space required in the node.
+
+An improvement over that last possibility is to use the CRI-O gRPC API, the
+equivalent of `circtl pull ...`. It doesn't improve performance or reduces the
+required size, but it means that there is one less component needed (no need
+for skopeo) and reduces the risks: CRI-O will be writing the images to the disk
+itself, so no risk of format mismatch.
+
+We aren't ruling out any of the above possibilities, consider them as
+implementation details, but we think that the last one is the better overall.
+
+#### CVO will need to orchestrate the upgrade activities
+
+To initiate the upgrade the administrator of the cluster changes the
+`ClusterVersion` object is changed like this:
+
+```yaml
+kind: ClusterVersion
+metadata:
+  name: version
+spec:
+  desiredUpdate:
+    version: 4.14.5
+```
+
+The cluster operator will first ask the machine config operator to pin and
+pre-load the release image, creating a `ContainerRuntimeConfig` object similar
+to this:
+
+```yaml
+apiVersion: machineconfiguration.openshift.io/v1
+kind: ContainerRuntimeConfig
+metadata:
+  name: pin-upgrade
+spec:
+  containerRuntimeConfig:
+    pinnedImages:
+    - quay.io/openshift-release-dev/ocp-release@sha256:...
+```
+
+The machine config operator will react to that pinning and pre-loading the
+image in all the nodes of the cluster, as described in the previous section.
+Once the image is pinned and pre-loaded the cluster version operator will
+inspect it and find out the references to the payload images. It will then also
+pin and pre-load those images, updating the `ContainerRuntimeConfig` object:
```

Comment:
It seems to me that this inspection of the release image is something that belongs into CVO more than MCO. I assume CVO already does something similar today to verify signatures, for example.

Reply:
I think the CVO should just act on the image not care about how the image got there. Certainly if the CVO is explicitly pulling the image we should fix that.

Reply:
CVO does not pull anything, because we know that image access is fiddly.  We get access to incoming release-image manifests by launching a `version-*` pod with a shared volume, so CRI-O can go fetch the image for us, the `version-*` containers copy the manifests into the shared volume, and the outgoing CVO grabs the manifests from the shared volume to start rolling out the update.